### ISSUE TITLE
Very slow development environment

### ISSUE THREAD

**Issue by Gmulti** [Shopify/shopify-demo-app-node-react#4]
# Issue summary

I just tested your demo. It works more or less but above all... the development environment is extremely slow. Do you have any advice on how to improve this ?

**Comment by loftadev** [comment]
I noticed this as well, I believe it is the Ngrok connection that is the cause.

**Comment by gfscott** [comment]
Thanks for letting us know about this — we’re looking into potential causes!

**Comment by nwtn** [comment]
Can you provide some extra detail about which parts of the dev process are slow (e.g. app load times, etc.)? Thanks

**Comment by aaronmarkle** [comment]
For me, when loading the app inside the test store that had the app installed was slow to load. Watching the ngrok terminal led me to believe ngrok was the issue. I would see requests for files come in on each load, but they were very slow to be served back.

**Comment by Popesites** [comment]
Can we get an update on this? I am experienced extremely laggy loadtime. I'm assuming it's because of ngrok being limited in resources.

**Comment by nwtn** [comment]
We haven’t been able to reproduce the slowness described here in our dev environments. At least, no more slowness than we might expect for a real roundtrip over the internet, which is what happens with ngrok.

Of course, just because we aren’t seeing the slowness doesn’t mean it’s not happening. If you suspect ngrok is the cause of the slow down, you can of course use other tunnelling services, like Serveo or Pagekit. If you end up finding that one of these speeds up your development, please let us know!

**Comment by tim-phillips** [comment]
I can confirm the slowness but I don't think there's enough evidence to blame ngrok, I've been using ngrok for some time now without issue and I don't get slowness on my other projects. I also get the same slowness with serveo. 

This happens on a fresh clone and npm install of this example package. Sometimes it takes over a minute but usually around 15-40 seconds. 

<img width="1280" alt="Screen Shot 2019-06-19 at 11 06 48 AM" src="https://user-images.githubusercontent.com/3578574/59789221-60443f80-9282-11e9-94c4-460208340198.png">

**Comment by uri3000** [comment]
Hi all, 
It is not related to ngrok. I believe its related to the next.js constantly recompiling the entire pages but not sure though. On production environment it is working fine, only during development its slow AS HELL (3+ minutes per page load).

@nwtn It is pretty much impossible to work this way. It get up to 3 minutes per page load and even more... I've been trying to work this way during the last few months and it has been a very painful experience.  My co-founder experiences the same thing and our 2 developers could barely work as well (same issue). We tried EVERYTHING - including different tunnel services, different next.js versions, react versions, polaris, everything...

We all used this codebase as a baseline for the project.

I BEG YOU guys, please have a look at this. I am willing to do a live demonstration for you at any time. My Skype is: abramson.uri, hangouts: abramson.uri at gmail.

Thank you.

**Comment by katiedavis** [comment]
hey @uri3000 I'm so sorry to hear about your issues with Next.js. This guide was created to give an example of how you can use different Shopify libraries and make an app for the Shopify ecosystem.

If the hot-module reloading is super slow for you, you don't need to use next as a framework. We used it for this tutorial simply because we wanted to teach Polaris, App Bridge and using our APIs, and didn't want to focus on webpack configuration or scaffolding an app. You can use the server you've already created and make a react app.

I did a quick search (I have not used the resources) and this is what I found:
https://github.com/JasonBoy/koa-web-kit
https://github.com/tokenfoundry/koa-react-ssr-example
https://medium.com/@s4y.solutions/server-side-rendering-with-react-another-guide-375883d86959

Hopefully they can be of use to you. We have tried to reproduce locally and do not have the same issues with slow load times, but I understand it can be super frustrating. You could also try reaching out the next.js community for help with slow load times! They might have a better idea of why it's loading so slowly: https://spectrum.chat/next-js

**Comment by uri3000** [comment]
Hi Katie and thank you for you response.
We have managed to locate the issue and cause for slowliness and it is
definitely the next.js recompiling on every single page load. If it is not
reproduced on your end it is probably because you are not using npm run dev
to run it locally - am I correct?

I have to say that if Shopify are putting up an example for how an app
should be built, you should know that a looot of developers are going to
count on your recommendations.

If this is not a recommended stack as you say, then you should write that
in bold large letters in even more than one place throughout the tutorial.

I understand the requirement to set up a quick and dirty demonstration of
an app and api usage, however, you should know that it is not how
developers perceive it. You have a lot of responsibility in that sense and
many developers around the world are counting on these app examples to be
the right selection of tech stack and use it to bootstrap their own project
- simply put, their entire product and business rely on this.

You can say that it is a generic api but eventually people prefer to start
with The Official example, thats just how it works.

I can merely ask or recommend you to change / fix, and it is not my
decision after all, but know that new developers who will start off an app
using this sample project will have an extremely slow and painful
development experience and this could end up being their impression of the
whole platform which is really too bad.

Just my 2 cents as a developer...
Thanks again for your time and response.

**Comment by pieperz** [comment]
I'm not sure how related this is, but I had similar issues with the Shopify-App-Cli and Nodemon.

Upgrading to [v1.19.1 ](https://github.com/remy/nodemon/releases/tag/v1.19.1) seemed to fix the issue.

**Comment by CelesteComet** [comment]
Any update on this yet? Development with NextJS is near impossible...

**Comment by uri3000** [comment]
Guys try running next not in development mode and the hot recompile will
not run. It makes it much faster. In addition try connecting through you
cell's hotspot, I know it sounds weird but give it a try

**Comment by CelesteComet** [comment]
This is pretty sad state... can we just get rid of this tutorial and replace it with an older one?

**Comment by mtonk** [comment]
I ran into this when starting out yesterday and found that if I upgrade next to the 9.x release (**npm i next@latest**), hot-module reloading is significantly faster.  I had to make a small change to _app.js to get rid of a deprecation warning, but everything still seems to be working.
